keepassxc-browser/content/observer-helper.js

```diff
@@ -0,0 +1,271 @@
+    const inputs = [];
+    for (const field of inputFields) {
+        if ((!ignoreVisibility && !kpxcFields.isVisible(field))
+            || kpxcFields.isSearchField(field)) {
+            continue;
+        }
+
+        const type = field.getLowerCaseAttribute('type');
+        if (kpxcObserverHelper.inputTypes.includes(type)) {
+            inputs.push(field);
+        }
+    }
+
+    return inputs;
+};
+
+// Checks if the input field has already identified at page load
+kpxcObserverHelper.alreadyIdentified = function(target) {
+    return kpxc.inputs.some(e => e === target);
+};
+
+// Adds elements to a monitor array. Identifies the input fields.
+kpxcObserverHelper.handleObserverAdd = async function(target) {
+    if (kpxcObserverHelper.ignoredElement(target)) {
+        return;
+    }
+
+    // Sometimes the settings haven't been loaded before new input fields are detected
+    if (Object.keys(kpxc.settings).length === 0) {
+        kpxc.init();
+        return;
+    }
+
+    const inputs = kpxcObserverHelper.getInputs(target);
+    if (inputs.length === 0) {
+        return;
+    }
+
+    await kpxc.initCombinations(inputs);
+    await kpxcIcons.initIcons(kpxc.combinations);
+
+    if (kpxc.databaseState === DatabaseState.UNLOCKED) {
+        if (_called.retrieveCredentials === false) {
+            await kpxc.retrieveCredentials();
+            return;
+        }
+
+        kpxc.prepareCredentials();
+    }
+};
+
+// Removes monitored elements
+kpxcObserverHelper.handleObserverRemove = function(target) {
+    if (kpxcObserverHelper.ignoredElement(target)) {
+        return;
+    }
+
+    const inputs = kpxcObserverHelper.getInputs(target, true);
+    if (inputs.length === 0) {
+        return;
+    }
+
+    kpxcIcons.deleteHiddenIcons();
+};
+
+// Handles CSS transitionend event
+kpxcObserverHelper.handleTransitionEnd = function(e) {
+    if (!e.isTrusted) {
+        return;
+    }
+
+    kpxcObserverHelper.handleObserverAdd(e.currentTarget);
+};
+
+// Returns true if element should be ignored
+kpxcObserverHelper.ignoredElement = function(target) {
+    if (kpxcObserverHelper.ignoredNode(target)) {
+        return true;
+    }
+
+    // Ignore elements that do not have a className (including SVG)
+    if (typeof target.className !== 'string') {
+        return true;
+    }
+
+    return false;
+};
+
+// Ignores all nodes that doesn't contain elements
+// Also ignore few Youtube-specific custom nodeNames
+kpxcObserverHelper.ignoredNode = function(target) {
+    if (!target
+        || kpxcObserverHelper.ignoredNodeTypes.some(e => e === target.nodeType)
+        || kpxcObserverHelper.ignoredNodeNames.some(e => e === target.nodeName)
+        || target.nodeName.startsWith('YTMUSIC')
+        || target.nodeName.startsWith('YT-')) {
+        return true;
+    }
+
+    return false;
+};
+
+// Initializes MutationObserver
+kpxcObserverHelper.initObserver = async function() {
```

Comment:
Recommend moving this to the top of the function definitions since it is the entry point. Generally speaking I prefer initializers and bootstrap functions near the top of the file for reading flow.